### ISSUE TITLE
[Bounty $1,500] [sdpa_decode] fp32_dest_acc_en=True silently produces wrong results for most shapes, and hangs at 96 query heads

### ISSUE THREAD

**Issue by kinginu** [tenstorrent/tt-metal#50564]
### Component / Area

`ops / transformer` — `ttnn.transformer.scaled_dot_product_attention_decode`

### Issue Type

Bad Outputs

### Observed

`scaled_dot_product_attention_decode` silently returns wrong results when the padded query-head count is above 32 (`PNHt > 1`) and `fp32_dest_acc_en=True`. Plain causal decode, no attention mask, no paging: with 64 query heads / 8 KV heads / head_dim 128, the worst per-head PCC is **0.359** instead of 0.999. 32 query heads is fine with either setting, and 64 query heads with `fp32_dest_acc_en=False` is fine, so the trigger is the flag together with `PNHt > 1`. It also needs more than one core per head-batch — `max_cores_per_head_batch=1` gives correct results.

With an **odd** `PNHt` (96 query heads) the same configuration **hangs the device** instead of returning wrong data; recovery needs `tt-smi -r`.

`fp32_dest_acc_en` halves DEST (`dst_size = fp32_dest_acc_en ? 4 : 8`, `sdpa_decode_program_factory.cpp:388`), and that value feeds the QK subblock derivation on the next line (`qk_out_subblock_h = min(PNHt, dst_size / qk_out_subblock_w)`) and `MUL_BCAST_GRANULARITY` at `:782`. Both are functions of `PNHt`, so they only change behaviour once `PNHt > 1`. #48748 already records the same class of corruption for Gemma4's `head_dim=512` decode config; this repro shows it is not head_dim-specific.

This combination looks untested: every sdpa decode unit test hardcodes `fp32_dest_acc_en=False`, and none uses more than 32 query heads.

### Expected

Either correct results with `fp32_dest_acc_en=True`, or reject the combination in `validate()`. Silently applying a wrong attention output is the worst of the three outcomes.

### 1. Steps (exact commands)

```bash
cd $TT_METAL_HOME && source python_env/bin/activate
python repro_sdpa_decode_fp32_dest_acc.py          # wrong results
python repro_sdpa_decode_fp32_dest_acc.py hang     # WARNING: hangs the device, needs tt-smi -r
```

<details><summary>repro_sdpa_decode_fp32_dest_acc.py (self-contained)</summary>

```python
# SPDX-FileCopyrightText: © 2026 Tenstorrent AI ULC
#
# SPDX-License-Identifier: Apache-2.0

"""Repro: sdpa_decode returns silently wrong results with fp32_dest_acc_en=True
when the padded query-head count exceeds 32 (PNHt > 1)."""
import math
import sys

import torch

import ttnn

HD, B, S, NKV, CUR_POS = 128, 1, 256, 8, 200


def pcc(a, b):
    a, b = a.flatten().double(), b.flatten().double()
    va, vb = a - a.mean(), b - b.mean()
    return float((va @ vb) / (va.norm() * vb.norm()))


def run(device, num_q_heads, fp32_dest_acc, max_cores=16):
    torch.manual_seed(0)
    k = torch.randn(B, NKV, S, HD).bfloat16().float()
    v = torch.randn(B, NKV, S, HD).bfloat16().float()
    q = torch.randn(1, B, num_q_heads, HD).bfloat16().float()
    cur_pos = torch.tensor([CUR_POS] * B, dtype=torch.int32)
    scale = 1.0 / math.sqrt(HD)

    tt = lambda t, dt: ttnn.Tensor(t, dt).to(ttnn.TILE_LAYOUT).to(device)
    out = ttnn.transformer.scaled_dot_product_attention_decode(
        tt(q, ttnn.bfloat16),
        tt(k, ttnn.bfloat16),
        tt(v, ttnn.bfloat16),
        cur_pos_tensor=ttnn.Tensor(cur_pos, ttnn.int32).to(device),
        is_causal=True,
        scale=scale,
        program_config=ttnn.SDPAProgramConfig(
            compute_with_storage_grid_size=device.compute_with_storage_grid_size(),
            q_chunk_size=32,
            k_chunk_size=64,
            exp_approx_mode=False,
            max_cores_per_head_batch=max_cores,
        ),
        compute_kernel_config=ttnn.init_device_compute_kernel_config(
            device.arch(),
            math_fidelity=ttnn.MathFidelity.HiFi2,
            math_approx_mode=True,
            fp32_dest_acc_en=fp32_dest_acc,
            packer_l1_acc=False,
        ),
    )
    got = ttnn.to_torch(out)[0, 0, :num_q_heads, :].float()

    group = num_q_heads // NKV
    worst = min(
        pcc(
            torch.softmax((k[0, i // group, : CUR_POS + 1] @ q[0, 0, i]) * scale, dim=-1)
            @ v[0, i // group, : CUR_POS + 1],
            got[i],
        )
        for i in range(num_q_heads)
    )
    return worst


device = ttnn.open_device(device_id=0)
try:
    for nh in (32, 64):
        for fp32 in (False, True):
            worst = run(device, nh, fp32)
            print(
                f"q_heads={nh:3d} (PNHt={nh // 32})  fp32_dest_acc_en={str(fp32):5s}  "
                f"worst per-head PCC {worst:.6f}  {'ok' if worst > 0.99 else '<-- WRONG'}",
                flush=True,
            )
    if len(sys.argv) > 1 and sys.argv[1] == "hang":
        print("96 q heads (PNHt=3) + fp32_dest_acc_en=True -> expect device hang", flush=True)
        run(device, 96, True)
        print("did not hang", flush=True)
finally:
    ttnn.close_device(device)
```

</details>

### 2. Input data / link or description

None — the script generates its own seeded random Q/K/V.

### 3. Frequency

Always (deterministic, both the wrong results and the hang).

### System Details

**1. Software Versions**

```
OS: Ubuntu 24.04.3 LTS (kernel 6.17.0-35-generic)
tt-metal: 0d4af3ae6e2c9aa226aebbe84350a6e0540a24e6 (main, 2026-07-17)
Driver: TT-KMD 2.7.0
Firmware bundle: 19.7.0.0
```

**2. Hardware Details**

```
Product: Blackhole
Card/System: p300c (repro runs on a single chip, device_id=0)
```

### Is this a regression?

No

### Regression Details

```
First bad version:   n/a - this configuration has never worked
Last known good:     none
Git bisect status:   not bisected; established from history (see below)
```

More than 32 query heads only started producing correct results at all with #34113 (68459b0e25, 2025-12-10), which fixed #33440. The three expressions that make the fp32 path wrong are unchanged since then — `dst_size = fp32_dest_acc_en ? 4 : 8`, `qk_out_subblock_h = min(PNHt, dst_size / qk_out_subblock_w)` (and the `out_out_subblock_h` twin), and `MUL_BCAST_GRANULARITY = min(PNHt * Sk_chunk_t, dst_size)`. They date back to 1901aedf96 (2024-08-19, the move of SDPA to ttnn); #36440 only moved the granularity into a helper without changing the formula. So `fp32_dest_acc_en=True` with `PNHt > 1` has never worked, rather than having regressed.

### Logs & Diagnostics

```
q_heads= 32 (PNHt=1)  fp32_dest_acc_en=False  worst per-head PCC 0.999664  ok
q_heads= 32 (PNHt=1)  fp32_dest_acc_en=True   worst per-head PCC 0.999605  ok
q_heads= 64 (PNHt=2)  fp32_dest_acc_en=False  worst per-head PCC 0.999362  ok
q_heads= 64 (PNHt=2)  fp32_dest_acc_en=True   worst per-head PCC 0.359092  <-- WRONG
96 q heads (PNHt=3) + fp32_dest_acc_en=True -> expect device hang
<hangs here; no further output, device needs tt-smi -r>
```

**Comment by kinginu** [comment]
I'm working on a PR for this.

**Comment by kinginu** [comment]
Correction: I scoped this too narrowly. Sweeping `head_dim` x `q_heads` x `k_chunk` shows `fp32_dest_acc_en=True` is wrong for **most** shapes, not just above 32 query heads — including the simplest one (32 heads, head_dim 128, k_chunk 32, PNHt=1), which gives PCC 0.011. `fp32_dest_acc_en=False` is correct in all 12 cells.

| head_dim | q_heads | k_chunk | fp32=False | fp32=True |
|---|---|---|---|---|
| 128 | 32 | 32 | 0.999652 | **0.010701** |
| 128 | 32 | 64 | 0.999664 | 0.999605 |
| 128 | 32 | 128 | 0.999564 | 0.999591 |
| 128 | 64 | 32 | 0.999475 | **0.183994** |
| 128 | 64 | 64 | 0.999362 | **0.359092** |
| 128 | 64 | 128 | 0.999372 | **0.481717** |
| 256 | 32 | 32 | 0.999544 | **0.227926** |
| 256 | 32 | 64 | 0.999595 | **0.334011** |
| 256 | 32 | 128 | 0.999583 | **0.369320** |
| 256 | 64 | 32 | 0.999458 | **0.284701** |
| 256 | 64 | 64 | 0.999514 | **0.392524** |
| 256 | 64 | 128 | 0.999481 | **0.453830** |

(worst per-head PCC, plain causal decode, 8 KV heads, same repro as above with `head_dim`/`k_chunk` parameterised.)

So the query-head count is not the trigger. Note the two passing cells and the failing ones can have *identical* derived program-config values — e.g. head_dim 128 / 32 heads / k_chunk 32 derives the same subblock sizes and granularities for both settings, and only `fp32=True` is wrong — so this is not the subblock/granularity derivation in the program factory but the DEST configuration itself. That is consistent with #48748, which saw it at head_dim 512 with 32 query heads.

The 96-query-head hang still reproduces as described.

**Comment by kinginu** [comment]
Opened https://github.com/tenstorrent/tt-metal/pull/50598 with a fix for the wrong results.

Root cause is `correction_block` in the cross-core flash-decode reduction: it runs five tiles through DEST in one `tile_regs_acquire`, which only fits while DEST holds eight tiles. With `fp32_dest_acc_en=True` it holds four, so the worker sum is read back as garbage — which is why the attention weights summed to 4-47 instead of 1, and why `max_cores_per_head_batch=1` (no reduction) was always correct.

Note the PR does **not** close this issue: the 96-query-head hang is a separate defect and still reproduces with the fix applied.

**Comment by kinginu** [comment]
Root-caused the hang too, and folded the fix into https://github.com/tenstorrent/tt-metal/pull/50598 (which now covers both symptoms and closes this issue).

The QK/out matmuls set `subblock_h = min(PNHt, dst_size / subblock_w)`, and `matmul_blocks` loops `PNHt / subblock_h` times over `subblock_h` rows. When `subblock_h` doesn't divide PNHt the integer division truncates, so the loop covers fewer than PNHt rows, the QK CB is under-produced, and the consumer waits forever. fp32 caps the subblock at 2, so an odd PNHt (e.g. 3 = 96 query heads, k_chunk 64) covers 2 of 3 rows and hangs; bf16 (cap 4) divides 3, which is why it was fp32-only. The PR clamps the cap to the largest divisor of PNHt. 96 and 160 query heads now run at PCC 0.9995.

**Comment by jberkowitzTT** [comment]
## 📝 Background

This is a community-suggested bounty that we have reviewed & assigned to @kinginu 


## 💡 Problem to Solve

>fp32_dest_acc_en=True silently produces wrong results for most shapes, and hangs at 96 query heads


Note: This bounty is beholden to the [Tenstorrent Bounty Program Terms & Conditions](https://docs.tenstorrent.com/bounty_terms.html) & local tax requirements may apply

**Comment by kinginu** [comment]
@jberkowitzTT could you run /codeowners ping in #50598 ?

**Comment by jberkowitzTT** [comment]
Running CI workflows now, once they pass I'll loop in the code owners

**Comment by kinginu** [comment]
Thank you!

**Comment by jberkowitzTT** [comment]
Hi @kinginu - congratulations on completing this bounty, and thank you for the great work!

To initiate the payment process, please email us at bounties@tenstorrent.com with a link to this issue
